crates/btcio/src/reader/query.rs

```diff
@@ -258,16 +252,26 @@ async fn poll_for_new_blocks<R: Reader>(
     for fetch_height in scan_start_height..=client_height {
         match fetch_and_process_block(ctx, fetch_height, state, status_updates).await {
             Ok((blkid, evs)) => {
-                events.extend_from_slice(&evs);
+                // Frist handle events, i.e. send to csm
+                for ev in evs.iter() {
+                    handle_bitcoin_event(ev.clone(), ctx, event_submitter).await?;
+                }
 
+                // If there is checkpoint in blocks, then wait until that gets accepted or
+                // discarded i.e. if it is verified. If it is accepted, update filter config.
                 if let Some(checkpt) = find_checkpoint_in_events(&evs) {
                     // if we have a checkpoint in this block, update filterconfig based on this
-                    let chainstate = borsh::from_slice(checkpt.checkpoint().sidecar().chainstate())
-                        .expect("deserialize chainstate");
+                    let accepted = wait_until_checkpoint_accepted_or_rejected(ctx, checkpt).await?;
+
+                    if accepted {
+                        let chainstate =
+                            borsh::from_slice(checkpt.checkpoint().sidecar().chainstate())
+                                .expect("deserialize chainstate");
```

Comment:
This shouldn't panic if it's malformed, even if the proof is (somehow) valid.

While we're here we should probably make this more organized, so that here it feels like you're just directly (but fallibly) taking the decoded sidecar field(s) out of the checkpoint structure, instead of needing to call borsh yourself here (especially since we'd like to move away from borsh and this removes one more place where we depend on it directly).